{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "01b5618e-3948-4dee-be0f-6ca2c2fdc43f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pgtools import gff_parser\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a03618-0862-4f1e-b2cc-076780f9d940",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6ef7044-f456-4e10-a662-852908a63cd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_cvg = {\n",
    "\"Pneumococcus\":85.12,\n",
    "\"Klebsiella\":84.94,\n",
    "\"E. coli\":84.41,\n",
    "\"E. coli fragmented\": 81.40,\n",
    "\"E. coli contaminated\": 84.22\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1a1e761a-82e6-489d-8f54-9cf56c0a9e28",
   "metadata": {},
   "outputs": [],
   "source": [
    "def datasets_mapping(x):\n",
    "    if x.startswith(\"sim\"):\n",
    "        x= x.replace(\"sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1\", \"E. coli \")\n",
    "    else:\n",
    "        x= x.replace(\"_subset\", \"\")\n",
    "        x= x.replace(\"GPSC\", \"Pneumococcus\")\n",
    "        x = x.replace(\"klebsiella\", \"Klebsiella\")\n",
    "    return x.strip().replace(\"_\",\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "21bc8c53-4254-4bd5-9f16-131235bd8798",
   "metadata": {},
   "outputs": [],
   "source": [
    "summaries_dir=\"/home/pampuch/studia/magisterka/final_summaries/dataset_summaries/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1a33f6d8-f437-4e57-b9e2-553598205d25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sim_gr_1e-11_lr_1e-12_mu_1e-14_rep1.csv\n",
      "Index(['genome', 'genome_size', 'contigs_n', 'N50'], dtype='object')\n",
      "sim_gr_1e-12_lr_1e-12_mu_1e-14_rep1.csv\n",
      "Index(['genome', 'genome_size', 'contigs_n', 'N50'], dtype='object')\n",
      "GPSC_subset.csv\n",
      "Index(['genome', 'genome_size', 'contigs_n', 'N50'], dtype='object')\n",
      "klebsiella_subset.csv\n",
      "Index(['genome', 'genome_size', 'contigs_n', 'N50'], dtype='object')\n",
      "summary.csv\n",
      "sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1.csv\n",
      "Index(['genome', 'genome_size', 'contigs_n', 'N50'], dtype='object')\n",
      "sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1_fragmented.csv\n",
      "Index(['genome', 'genome_size', 'contigs_n', 'N50'], dtype='object')\n",
      "sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1_contaminated.csv\n",
      "Index(['genome', 'genome_size', 'contigs_n', 'N50'], dtype='object')\n",
      "[['simgr1e-11lr1e-12mu1e-14rep1', 107.07, 5.8070924938667225, 140561.0, 5801.96580409771], ['simgr1e-12lr1e-12mu1e-14rep1', 145.89, 4.485430961355989, 169733.79, 318.18060099334286], ['Pneumococcus', 124.6, 151.51529365224593, 115445.85, 46288.96020852966], ['Klebsiella', 780.4, 481.4431594585945, 74041.3, 60627.554601500255], ['E. coli', 123.37, 7.706294161071825, 130416.64, 1765.785529377103], ['E. coli fragmented', 322.54, 8.41765820742145, 23072.01, 1358.0759822765306], ['E. coli contaminated', 137.43, 6.779730634379767, 110721.06, 7555.8367415561925]]\n"
     ]
    }
   ],
   "source": [
    "columns = [\"Dataset\",\"Średnia liczba contigów w genomie\", \"Std liczby contigów w genomie\", \"Średnie N50 genomu\", \"Std N50 genomu\"]\n",
    "rows = []\n",
    "for f in os.listdir(summaries_dir):\n",
    "    print(f)\n",
    "    if f == \"summary.csv\":\n",
    "        continue\n",
    "    row = []\n",
    "    df = pd.read_csv(os.path.join(summaries_dir, f))\n",
    "    row.append(datasets_mapping(f.split(\".\")[0]))\n",
    "    print(df.columns)\n",
    "    row += [df[\"contigs_n\"].mean(), df[\"contigs_n\"].std(), df[\"N50\"].mean(), df[\"N50\"].std()]\n",
    "    rows.append(row)\n",
    "print(rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "08a2805b-4d99-47a0-b5cd-992634dbbe19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Średnia liczba contigów w genomie</th>\n",
       "      <th>Std liczby contigów w genomie</th>\n",
       "      <th>Średnie N50 genomu</th>\n",
       "      <th>Std N50 genomu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>123.37</td>\n",
       "      <td>7.71</td>\n",
       "      <td>130416.64</td>\n",
       "      <td>1765.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>137.43</td>\n",
       "      <td>6.78</td>\n",
       "      <td>110721.06</td>\n",
       "      <td>7555.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>322.54</td>\n",
       "      <td>8.42</td>\n",
       "      <td>23072.01</td>\n",
       "      <td>1358.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>780.40</td>\n",
       "      <td>481.44</td>\n",
       "      <td>74041.30</td>\n",
       "      <td>60627.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>124.60</td>\n",
       "      <td>151.52</td>\n",
       "      <td>115445.85</td>\n",
       "      <td>46288.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>simgr1e-11lr1e-12mu1e-14rep1</td>\n",
       "      <td>107.07</td>\n",
       "      <td>5.81</td>\n",
       "      <td>140561.00</td>\n",
       "      <td>5801.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>simgr1e-12lr1e-12mu1e-14rep1</td>\n",
       "      <td>145.89</td>\n",
       "      <td>4.49</td>\n",
       "      <td>169733.79</td>\n",
       "      <td>318.18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Dataset  Średnia liczba contigów w genomie  \\\n",
       "4                       E. coli                             123.37   \n",
       "6          E. coli contaminated                             137.43   \n",
       "5            E. coli fragmented                             322.54   \n",
       "3                    Klebsiella                             780.40   \n",
       "2                  Pneumococcus                             124.60   \n",
       "0  simgr1e-11lr1e-12mu1e-14rep1                             107.07   \n",
       "1  simgr1e-12lr1e-12mu1e-14rep1                             145.89   \n",
       "\n",
       "   Std liczby contigów w genomie  Średnie N50 genomu  Std N50 genomu  \n",
       "4                           7.71           130416.64         1765.79  \n",
       "6                           6.78           110721.06         7555.84  \n",
       "5                           8.42            23072.01         1358.08  \n",
       "3                         481.44            74041.30        60627.55  \n",
       "2                         151.52           115445.85        46288.96  \n",
       "0                           5.81           140561.00         5801.97  \n",
       "1                           4.49           169733.79          318.18  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_conts = pd.DataFrame(data=rows, columns=columns)\n",
    "df_conts = df_conts.sort_values(by=\"Dataset\")\n",
    "for col in df_conts.columns[1:]:\n",
    "    df_conts[col] = df_conts[col].round(2)\n",
    "df_conts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6ed591c9-c222-41d5-a649-0ee19afe84e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_conts.to_csv(os.path.join(summaries_dir, \"summary.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d5b8499-211f-4c03-92f9-9b03f5fd285c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Pneumococcus': 85.12,\n",
       " 'Klebsiella': 84.94,\n",
       " 'E. coli': 84.41,\n",
       " 'E. coli fragmented': 81.4,\n",
       " 'E. coli contaminated': 84.22}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_cvg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c22b7b13-d046-457d-a7ce-c722e14422b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_gffs_dir = \"/home/pampuch/studia/magisterka/test_data/gffs/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "27b5a41a-518e-4eed-841c-d74962f4aba7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1_fragmented\",\n",
      "\"sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1_contaminated\",\n",
      "\"klebsiella_subset\",\n",
      "\"sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1\",\n",
      "\"sim_gr_1e-12_lr_1e-12_mu_1e-14_rep1\",\n",
      "\"sim_gr_1e-11_lr_1e-12_mu_1e-14_rep1\",\n",
      "\"GPSC_subset\",\n"
     ]
    }
   ],
   "source": [
    "for dataset in os.listdir(all_gffs_dir):\n",
    "    print(f'\"{dataset}\",') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2f4336b8-4495-4eab-bb27-55463552bda8",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = [\n",
    "\"sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1\",\n",
    "\"sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1_contaminated\",\n",
    "\"sim_gr_1e-11_lr_1e-12_mu_1e-15_rep1_fragmented\",\n",
    "\"klebsiella_subset\",\n",
    "\"GPSC_subset\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e53a2fe1-2b9d-483e-ac36-bec81f264ab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "kleb_gff = os.path.join(all_gffs_dir, datasets[3], \"gff\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f59036f6-38e9-41e6-90bb-82d8146c0afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\n",
    "\"Dataset\",\n",
    "\"Liczba genomów\",\n",
    "\"Sumaryczna długość genomów\",\n",
    "\"Sumaryczna liczba kontigów\",\n",
    "\"Średnia długość kontigu\",\n",
    "\"Średni rozmiar genomu\",\n",
    "\"Średnia długość kontigu w genomie\",\n",
    "\"Średnia liczba kontigów w genomie\",\n",
    "\"Pokrycie genomów przez CDS (%)\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "1debf38f-3f75-4fd5-9868-e7754f29c4ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basic_stats(genome_gffs):\n",
    "    all_mean_cont_lens = []\n",
    "    # all_mean_cont_n = []\n",
    "    all_cont_lens = []\n",
    "    all_contig_n = []\n",
    "    all_genome_lens = []\n",
    "    genomes_n = 0\n",
    "    # all_len = 0\n",
    "    gff_all = gff_parser.parse_GFFs_dir(genome_gffs, gff_simple=False)\n",
    "    for genome, gff in gff_all.items():\n",
    "        cont_lens = []\n",
    "        genomes_n += 1\n",
    "        cont_n = 0\n",
    "        for scaff in gff.scaffolds:\n",
    "            cont_n += 1\n",
    "            cont_lens.append(scaff.length)\n",
    "            all_cont_lens.append(scaff.length)\n",
    "        all_genome_lens.append(sum(cont_lens))\n",
    "        all_contig_n.append(cont_n)\n",
    "        all_mean_cont_lens.append(sum(cont_lens) / cont_n)\n",
    "\n",
    "    res_dict = {\n",
    "        \"Liczba genomów\": genomes_n,\n",
    "        \"Sumaryczna długość genomów\": sum(all_genome_lens),\n",
    "        \"Sumaryczna liczba kontigów\": sum(all_contig_n),\n",
    "        \"Średnia długość kontigu\": sum(all_cont_lens) / len(all_cont_lens),\n",
    "        \"Średni rozmiar genomu\": sum(all_genome_lens) / genomes_n,\n",
    "        \"Średnia długość kontigu w genomie\": sum(all_mean_cont_lens) / genomes_n,\n",
    "        \"Średnia liczba kontigów w genomie\": sum(all_contig_n) / genomes_n,\n",
    "        # \"Pokrycie genomów przez CDS (%)\",\n",
    "    }\n",
    "\n",
    "    return res_dict\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "bab6d093-53c3-49f6-b6d4-b6aa124bd5e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dict = basic_stats(kleb_gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "403f9919-01e2-4f03-a66d-93ff9207a595",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Liczba genomów': 20,\n",
       " 'Sumaryczna długość genomów': 117375178,\n",
       " 'Sumaryczna liczba kontigów': 15608,\n",
       " 'Średni rozmiar genomu': 5868758.9,\n",
       " 'Średnia długość kontigu w genomie': 11384.184298403612,\n",
       " 'Średnia liczba kontigów w genomie': 780.4}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "9893825e-47a9-4041-8af8-3f06522dc20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_dicts = {}\n",
    "for dataset in datasets:\n",
    "    dir = os.path.join(all_gffs_dir, dataset, \"gff\")\n",
    "    summary_dicts[datasets_mapping(dataset)] = basic_stats(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bba892a5-6d93-4cdd-9acb-9aa41a7f0578",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'E. coli': {'Liczba genomów': 100,\n",
       "  'Sumaryczna długość genomów': 347207379,\n",
       "  'Sumaryczna liczba kontigów': 12337,\n",
       "  'Średni rozmiar genomu': 3472073.79,\n",
       "  'Średnia długość kontigu w genomie': 28250.17757423725,\n",
       "  'Średnia liczba kontigów w genomie': 123.37},\n",
       " 'E. coli contaminated': {'Liczba genomów': 100,\n",
       "  'Sumaryczna długość genomów': 348927115,\n",
       "  'Sumaryczna liczba kontigów': 13743,\n",
       "  'Średni rozmiar genomu': 3489271.15,\n",
       "  'Średnia długość kontigu w genomie': 25449.42886845667,\n",
       "  'Średnia liczba kontigów w genomie': 137.43},\n",
       " 'E. coli fragmented': {'Liczba genomów': 100,\n",
       "  'Sumaryczna długość genomów': 347126358,\n",
       "  'Sumaryczna liczba kontigów': 32254,\n",
       "  'Średni rozmiar genomu': 3471263.58,\n",
       "  'Średnia długość kontigu w genomie': 10769.495195841278,\n",
       "  'Średnia liczba kontigów w genomie': 322.54},\n",
       " 'Klebsiella': {'Liczba genomów': 20,\n",
       "  'Sumaryczna długość genomów': 117375178,\n",
       "  'Sumaryczna liczba kontigów': 15608,\n",
       "  'Średni rozmiar genomu': 5868758.9,\n",
       "  'Średnia długość kontigu w genomie': 11384.184298403612,\n",
       "  'Średnia liczba kontigów w genomie': 780.4},\n",
       " 'Pneumococcus': {'Liczba genomów': 20,\n",
       "  'Sumaryczna długość genomów': 45919085,\n",
       "  'Sumaryczna liczba kontigów': 2492,\n",
       "  'Średni rozmiar genomu': 2295954.25,\n",
       "  'Średnia długość kontigu w genomie': 36545.615865109445,\n",
       "  'Średnia liczba kontigów w genomie': 124.6}}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_dicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "15d614bf-81b3-414d-9b70-f8639f01d417",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in summary_dicts:\n",
    "    summary_dicts[k][\"Pokrycie genomów przez CDS (%)\"] =  cds_cvg[k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ec78a1d1-cdec-42d6-91b3-6f4a43ee11db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'E. coli': {'Liczba genomów': 100,\n",
       "  'Sumaryczna długość genomów': 347207379,\n",
       "  'Sumaryczna liczba kontigów': 12337,\n",
       "  'Średni rozmiar genomu': 3472073.79,\n",
       "  'Średnia długość kontigu w genomie': 28250.17757423725,\n",
       "  'Średnia liczba kontigów w genomie': 123.37,\n",
       "  'Pokrycie genomów przez CDS': 84.41},\n",
       " 'E. coli contaminated': {'Liczba genomów': 100,\n",
       "  'Sumaryczna długość genomów': 348927115,\n",
       "  'Sumaryczna liczba kontigów': 13743,\n",
       "  'Średni rozmiar genomu': 3489271.15,\n",
       "  'Średnia długość kontigu w genomie': 25449.42886845667,\n",
       "  'Średnia liczba kontigów w genomie': 137.43,\n",
       "  'Pokrycie genomów przez CDS': 84.22},\n",
       " 'E. coli fragmented': {'Liczba genomów': 100,\n",
       "  'Sumaryczna długość genomów': 347126358,\n",
       "  'Sumaryczna liczba kontigów': 32254,\n",
       "  'Średni rozmiar genomu': 3471263.58,\n",
       "  'Średnia długość kontigu w genomie': 10769.495195841278,\n",
       "  'Średnia liczba kontigów w genomie': 322.54,\n",
       "  'Pokrycie genomów przez CDS': 81.4},\n",
       " 'Klebsiella': {'Liczba genomów': 20,\n",
       "  'Sumaryczna długość genomów': 117375178,\n",
       "  'Sumaryczna liczba kontigów': 15608,\n",
       "  'Średni rozmiar genomu': 5868758.9,\n",
       "  'Średnia długość kontigu w genomie': 11384.184298403612,\n",
       "  'Średnia liczba kontigów w genomie': 780.4,\n",
       "  'Pokrycie genomów przez CDS': 84.94},\n",
       " 'Pneumococcus': {'Liczba genomów': 20,\n",
       "  'Sumaryczna długość genomów': 45919085,\n",
       "  'Sumaryczna liczba kontigów': 2492,\n",
       "  'Średni rozmiar genomu': 2295954.25,\n",
       "  'Średnia długość kontigu w genomie': 36545.615865109445,\n",
       "  'Średnia liczba kontigów w genomie': 124.6,\n",
       "  'Pokrycie genomów przez CDS': 85.12}}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_dicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "13aa0d65-824b-4aa3-a688-02df33455383",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Dataset',\n",
       " 'Liczba genomów',\n",
       " 'Sumaryczna długość genomów',\n",
       " 'Sumaryczna liczba kontigów',\n",
       " 'Średni rozmiar genomu',\n",
       " 'Średnia długość kontigu w genomie',\n",
       " 'Średnia liczba kontigów w genomie',\n",
       " 'Pokrycie genomów przez CDS (%)']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0ba66276-2703-4b9f-9d26-a1280745cfa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_row_vals = []\n",
    "for dataset, vals in summary_dicts.items():\n",
    "    row_vals = [dataset]\n",
    "    for k in columns[1:]:\n",
    "        row_vals.append(vals[k])\n",
    "    all_row_vals.append(row_vals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e484f46d-9361-435b-ba48-a6769619d937",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['E. coli',\n",
       "  100,\n",
       "  347207379,\n",
       "  12337,\n",
       "  3472073.79,\n",
       "  28250.17757423725,\n",
       "  123.37,\n",
       "  84.41],\n",
       " ['E. coli contaminated',\n",
       "  100,\n",
       "  348927115,\n",
       "  13743,\n",
       "  3489271.15,\n",
       "  25449.42886845667,\n",
       "  137.43,\n",
       "  84.22],\n",
       " ['E. coli fragmented',\n",
       "  100,\n",
       "  347126358,\n",
       "  32254,\n",
       "  3471263.58,\n",
       "  10769.495195841278,\n",
       "  322.54,\n",
       "  81.4],\n",
       " ['Klebsiella',\n",
       "  20,\n",
       "  117375178,\n",
       "  15608,\n",
       "  5868758.9,\n",
       "  11384.184298403612,\n",
       "  780.4,\n",
       "  84.94],\n",
       " ['Pneumococcus',\n",
       "  20,\n",
       "  45919085,\n",
       "  2492,\n",
       "  2295954.25,\n",
       "  36545.615865109445,\n",
       "  124.6,\n",
       "  85.12]]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_row_vals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4f9ab3d2-fbac-4524-8d6a-6817209df3e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dataset = pd.DataFrame(columns=columns, data=all_row_vals)\n",
    "for col in df_dataset.columns:\n",
    "    if col.startswith(\"Ś\"):\n",
    "        df_dataset[col] = df_dataset[col].round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "7b77c242-9265-4dcf-b67b-f8d8af45deac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\"Dataset\",\n",
      "\"Liczba genomów\",\n",
      "\"Sumaryczna długość genomów\",\n",
      "\"Sumaryczna liczba kontigów\",\n",
      "\"Średnia długość kontigu\",\n",
      "\"Średni rozmiar genomu\",\n",
      "\"Średnia długość kontigu w genomie\",\n",
      "\"Średnia liczba kontigów w genomie\",\n",
      "\"Pokrycie genomów przez CDS (%)\",\n"
     ]
    }
   ],
   "source": [
    "for col in df_dataset.columns:\n",
    "    print(f'\"{col}\",')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "81d8519a-4f52-4cd8-8f56-8f0c6f06839e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dataset = df_dataset[[\"Dataset\",\n",
    "\"Liczba genomów\",\n",
    "\"Sumaryczna długość genomów\",\n",
    "\"Średni rozmiar genomu\",\n",
    "\"Pokrycie genomów przez CDS (%)\",\n",
    "\"Sumaryczna liczba kontigów\",\n",
    "\"Średnia liczba kontigów w genomie\",\n",
    "\"Średnia długość kontigu\",\n",
    "\"Średnia długość kontigu w genomie\",\n",
    "]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "daf72f29-1177-4b9f-be07-ff2572e04a9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Liczba genomów</th>\n",
       "      <th>Sumaryczna długość genomów</th>\n",
       "      <th>Średni rozmiar genomu</th>\n",
       "      <th>Pokrycie genomów przez CDS (%)</th>\n",
       "      <th>Sumaryczna liczba kontigów</th>\n",
       "      <th>Średnia liczba kontigów w genomie</th>\n",
       "      <th>Średnia długość kontigu</th>\n",
       "      <th>Średnia długość kontigu w genomie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>100</td>\n",
       "      <td>347207379</td>\n",
       "      <td>3472073.79</td>\n",
       "      <td>84.41</td>\n",
       "      <td>12337</td>\n",
       "      <td>123.37</td>\n",
       "      <td>28143.58</td>\n",
       "      <td>28250.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>100</td>\n",
       "      <td>348927115</td>\n",
       "      <td>3489271.15</td>\n",
       "      <td>84.22</td>\n",
       "      <td>13743</td>\n",
       "      <td>137.43</td>\n",
       "      <td>25389.44</td>\n",
       "      <td>25449.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>100</td>\n",
       "      <td>347126358</td>\n",
       "      <td>3471263.58</td>\n",
       "      <td>81.40</td>\n",
       "      <td>32254</td>\n",
       "      <td>322.54</td>\n",
       "      <td>10762.27</td>\n",
       "      <td>10769.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>20</td>\n",
       "      <td>117375178</td>\n",
       "      <td>5868758.90</td>\n",
       "      <td>84.94</td>\n",
       "      <td>15608</td>\n",
       "      <td>780.40</td>\n",
       "      <td>7520.19</td>\n",
       "      <td>11384.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>20</td>\n",
       "      <td>45919085</td>\n",
       "      <td>2295954.25</td>\n",
       "      <td>85.12</td>\n",
       "      <td>2492</td>\n",
       "      <td>124.60</td>\n",
       "      <td>18426.60</td>\n",
       "      <td>36545.62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Dataset  Liczba genomów  Sumaryczna długość genomów  \\\n",
       "0               E. coli             100                   347207379   \n",
       "1  E. coli contaminated             100                   348927115   \n",
       "2    E. coli fragmented             100                   347126358   \n",
       "3            Klebsiella              20                   117375178   \n",
       "4          Pneumococcus              20                    45919085   \n",
       "\n",
       "   Średni rozmiar genomu  Pokrycie genomów przez CDS (%)  \\\n",
       "0             3472073.79                           84.41   \n",
       "1             3489271.15                           84.22   \n",
       "2             3471263.58                           81.40   \n",
       "3             5868758.90                           84.94   \n",
       "4             2295954.25                           85.12   \n",
       "\n",
       "   Sumaryczna liczba kontigów  Średnia liczba kontigów w genomie  \\\n",
       "0                       12337                             123.37   \n",
       "1                       13743                             137.43   \n",
       "2                       32254                             322.54   \n",
       "3                       15608                             780.40   \n",
       "4                        2492                             124.60   \n",
       "\n",
       "   Średnia długość kontigu  Średnia długość kontigu w genomie  \n",
       "0                 28143.58                           28250.18  \n",
       "1                 25389.44                           25449.43  \n",
       "2                 10762.27                           10769.50  \n",
       "3                  7520.19                           11384.18  \n",
       "4                 18426.60                           36545.62  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "53b54b7e-3620-4496-be5d-0ff85bfe9f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dataset.to_csv(\"/home/pampuch/studia/magisterka/final_summaries/basic_dataset_stats.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "7d55c5bb-d1e5-4e71-8783-b3ef90cc9cf5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    28250.18\n",
       "1    25449.43\n",
       "2    10769.50\n",
       "3    11384.18\n",
       "4    36545.62\n",
       "Name: Średnia długość kontigu w genomie, dtype: float64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dataset[\"Średnia długość kontigu w genomie\"].round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "a485bf83-5958-4cae-9b26-5098c4603ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stats = pd.read_csv(\"/home/pampuch/studia/magisterka/final_summaries/stats.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "c2bfcbc2-07cf-402b-9e26-594195f8abe7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Model</th>\n",
       "      <th>Udział przerw (%)</th>\n",
       "      <th>Liczba bloków</th>\n",
       "      <th>Łączna długość zawartych sekwencji</th>\n",
       "      <th>Średni rozmiar bloku</th>\n",
       "      <th>Średnia długość sekwencji w bloku</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>0.11</td>\n",
       "      <td>4234</td>\n",
       "      <td>292139716</td>\n",
       "      <td>72.90</td>\n",
       "      <td>945.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>62.29</td>\n",
       "      <td>2246</td>\n",
       "      <td>83733481</td>\n",
       "      <td>66.17</td>\n",
       "      <td>213.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>13.96</td>\n",
       "      <td>6480</td>\n",
       "      <td>375873197</td>\n",
       "      <td>70.57</td>\n",
       "      <td>707.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>1.48</td>\n",
       "      <td>1333</td>\n",
       "      <td>332387749</td>\n",
       "      <td>92.35</td>\n",
       "      <td>2629.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>0.25</td>\n",
       "      <td>4247</td>\n",
       "      <td>292302476</td>\n",
       "      <td>72.62</td>\n",
       "      <td>945.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>62.33</td>\n",
       "      <td>2251</td>\n",
       "      <td>82821561</td>\n",
       "      <td>65.82</td>\n",
       "      <td>211.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>13.96</td>\n",
       "      <td>6498</td>\n",
       "      <td>375124037</td>\n",
       "      <td>70.27</td>\n",
       "      <td>707.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>1.33</td>\n",
       "      <td>1550</td>\n",
       "      <td>328136320</td>\n",
       "      <td>90.22</td>\n",
       "      <td>2287.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>4.72</td>\n",
       "      <td>4256</td>\n",
       "      <td>294386131</td>\n",
       "      <td>72.12</td>\n",
       "      <td>913.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>65.27</td>\n",
       "      <td>2337</td>\n",
       "      <td>84756187</td>\n",
       "      <td>59.98</td>\n",
       "      <td>211.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>18.25</td>\n",
       "      <td>6593</td>\n",
       "      <td>379142318</td>\n",
       "      <td>67.81</td>\n",
       "      <td>693.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>6.80</td>\n",
       "      <td>6090</td>\n",
       "      <td>157744627</td>\n",
       "      <td>45.38</td>\n",
       "      <td>515.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>2.29</td>\n",
       "      <td>7363</td>\n",
       "      <td>342178554</td>\n",
       "      <td>98.60</td>\n",
       "      <td>460.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>5.09</td>\n",
       "      <td>10304</td>\n",
       "      <td>101075318</td>\n",
       "      <td>10.25</td>\n",
       "      <td>908.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>48.78</td>\n",
       "      <td>4512</td>\n",
       "      <td>16982123</td>\n",
       "      <td>10.83</td>\n",
       "      <td>179.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>11.37</td>\n",
       "      <td>14816</td>\n",
       "      <td>118057441</td>\n",
       "      <td>10.42</td>\n",
       "      <td>677.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>7.03</td>\n",
       "      <td>4021</td>\n",
       "      <td>33443360</td>\n",
       "      <td>4.23</td>\n",
       "      <td>1444.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>6.24</td>\n",
       "      <td>7844</td>\n",
       "      <td>104370735</td>\n",
       "      <td>17.09</td>\n",
       "      <td>729.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>4.95</td>\n",
       "      <td>3323</td>\n",
       "      <td>40604368</td>\n",
       "      <td>13.64</td>\n",
       "      <td>851.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>50.27</td>\n",
       "      <td>2194</td>\n",
       "      <td>22531171</td>\n",
       "      <td>12.58</td>\n",
       "      <td>406.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>21.13</td>\n",
       "      <td>5517</td>\n",
       "      <td>63135539</td>\n",
       "      <td>13.22</td>\n",
       "      <td>683.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>3.78</td>\n",
       "      <td>1504</td>\n",
       "      <td>19601305</td>\n",
       "      <td>7.78</td>\n",
       "      <td>1612.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>6.54</td>\n",
       "      <td>2176</td>\n",
       "      <td>41762422</td>\n",
       "      <td>16.99</td>\n",
       "      <td>1055.46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Dataset                    Model  Udział przerw (%)  \\\n",
       "0                E. coli       Panaroo podstawowy               0.11   \n",
       "1                E. coli     Uzupełnienie Panaroo              62.29   \n",
       "2                E. coli  Panaroo z uzupełnieniem              13.96   \n",
       "3                E. coli                   Cactus               1.48   \n",
       "4   E. coli contaminated       Panaroo podstawowy               0.25   \n",
       "5   E. coli contaminated     Uzupełnienie Panaroo              62.33   \n",
       "6   E. coli contaminated  Panaroo z uzupełnieniem              13.96   \n",
       "7   E. coli contaminated                   Cactus               1.33   \n",
       "8     E. coli fragmented       Panaroo podstawowy               4.72   \n",
       "9     E. coli fragmented     Uzupełnienie Panaroo              65.27   \n",
       "10    E. coli fragmented  Panaroo z uzupełnieniem              18.25   \n",
       "11    E. coli fragmented                   Cactus               6.80   \n",
       "12    E. coli fragmented              Progressive               2.29   \n",
       "13            Klebsiella       Panaroo podstawowy               5.09   \n",
       "14            Klebsiella     Uzupełnienie Panaroo              48.78   \n",
       "15            Klebsiella  Panaroo z uzupełnieniem              11.37   \n",
       "16            Klebsiella                   Cactus               7.03   \n",
       "17            Klebsiella              Progressive               6.24   \n",
       "18          Pneumococcus       Panaroo podstawowy               4.95   \n",
       "19          Pneumococcus     Uzupełnienie Panaroo              50.27   \n",
       "20          Pneumococcus  Panaroo z uzupełnieniem              21.13   \n",
       "21          Pneumococcus                   Cactus               3.78   \n",
       "22          Pneumococcus              Progressive               6.54   \n",
       "\n",
       "    Liczba bloków  Łączna długość zawartych sekwencji  Średni rozmiar bloku  \\\n",
       "0            4234                           292139716                 72.90   \n",
       "1            2246                            83733481                 66.17   \n",
       "2            6480                           375873197                 70.57   \n",
       "3            1333                           332387749                 92.35   \n",
       "4            4247                           292302476                 72.62   \n",
       "5            2251                            82821561                 65.82   \n",
       "6            6498                           375124037                 70.27   \n",
       "7            1550                           328136320                 90.22   \n",
       "8            4256                           294386131                 72.12   \n",
       "9            2337                            84756187                 59.98   \n",
       "10           6593                           379142318                 67.81   \n",
       "11           6090                           157744627                 45.38   \n",
       "12           7363                           342178554                 98.60   \n",
       "13          10304                           101075318                 10.25   \n",
       "14           4512                            16982123                 10.83   \n",
       "15          14816                           118057441                 10.42   \n",
       "16           4021                            33443360                  4.23   \n",
       "17           7844                           104370735                 17.09   \n",
       "18           3323                            40604368                 13.64   \n",
       "19           2194                            22531171                 12.58   \n",
       "20           5517                            63135539                 13.22   \n",
       "21           1504                            19601305                  7.78   \n",
       "22           2176                            41762422                 16.99   \n",
       "\n",
       "    Średnia długość sekwencji w bloku  \n",
       "0                              945.35  \n",
       "1                              213.45  \n",
       "2                              707.48  \n",
       "3                             2629.29  \n",
       "4                              945.38  \n",
       "5                              211.56  \n",
       "6                              707.26  \n",
       "7                             2287.57  \n",
       "8                              913.90  \n",
       "9                              211.02  \n",
       "10                             693.54  \n",
       "11                             515.54  \n",
       "12                             460.53  \n",
       "13                             908.55  \n",
       "14                             179.06  \n",
       "15                             677.81  \n",
       "16                            1444.11  \n",
       "17                             729.90  \n",
       "18                             851.21  \n",
       "19                             406.89  \n",
       "20                             683.07  \n",
       "21                            1612.74  \n",
       "22                            1055.46  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "58b3ac4c-1551-4659-8a7e-410e1710ab7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stats = df_stats.join(df_dataset.set_index(\"Dataset\")[\"Sumaryczna długość genomów\"], on=\"Dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "9959b024-086e-4748-bc72-5114553dfd5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stats[\"Pokrycie pangenomu\"] = df_stats[\"Łączna długość zawartych sekwencji\"] * (1 - df_stats[\"Udział przerw (%)\"]/100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "2d0d23b4-d83e-4a4a-adc3-cb30c3ffbfc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stats[\"Pokrycie pangenomu (%)\"] = df_stats[\"Pokrycie pangenomu\"] * 100 /df_stats[\"Sumaryczna długość genomów\"]\n",
    "df_stats[\"Pokrycie pangenomu (%)\"] = df_stats[\"Pokrycie pangenomu (%)\"].round(2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "d2bb13ad-1336-4d85-89de-ad943b457468",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Model</th>\n",
       "      <th>Udział przerw (%)</th>\n",
       "      <th>Liczba bloków</th>\n",
       "      <th>Łączna długość zawartych sekwencji</th>\n",
       "      <th>Średni rozmiar bloku</th>\n",
       "      <th>Średnia długość sekwencji w bloku</th>\n",
       "      <th>Sumaryczna długość genomów</th>\n",
       "      <th>Pokrycie pangenomu</th>\n",
       "      <th>Pokrycie pangenomu (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>0.11</td>\n",
       "      <td>4234</td>\n",
       "      <td>292139716</td>\n",
       "      <td>72.90</td>\n",
       "      <td>945.35</td>\n",
       "      <td>347207379</td>\n",
       "      <td>2.918184e+08</td>\n",
       "      <td>84.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>62.29</td>\n",
       "      <td>2246</td>\n",
       "      <td>83733481</td>\n",
       "      <td>66.17</td>\n",
       "      <td>213.45</td>\n",
       "      <td>347207379</td>\n",
       "      <td>3.157590e+07</td>\n",
       "      <td>9.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>13.96</td>\n",
       "      <td>6480</td>\n",
       "      <td>375873197</td>\n",
       "      <td>70.57</td>\n",
       "      <td>707.48</td>\n",
       "      <td>347207379</td>\n",
       "      <td>3.234013e+08</td>\n",
       "      <td>93.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E. coli</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>1.48</td>\n",
       "      <td>1333</td>\n",
       "      <td>332387749</td>\n",
       "      <td>92.35</td>\n",
       "      <td>2629.29</td>\n",
       "      <td>347207379</td>\n",
       "      <td>3.274684e+08</td>\n",
       "      <td>94.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>0.25</td>\n",
       "      <td>4247</td>\n",
       "      <td>292302476</td>\n",
       "      <td>72.62</td>\n",
       "      <td>945.38</td>\n",
       "      <td>348927115</td>\n",
       "      <td>2.915717e+08</td>\n",
       "      <td>83.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>62.33</td>\n",
       "      <td>2251</td>\n",
       "      <td>82821561</td>\n",
       "      <td>65.82</td>\n",
       "      <td>211.56</td>\n",
       "      <td>348927115</td>\n",
       "      <td>3.119888e+07</td>\n",
       "      <td>8.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>13.96</td>\n",
       "      <td>6498</td>\n",
       "      <td>375124037</td>\n",
       "      <td>70.27</td>\n",
       "      <td>707.26</td>\n",
       "      <td>348927115</td>\n",
       "      <td>3.227567e+08</td>\n",
       "      <td>92.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>E. coli contaminated</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>1.33</td>\n",
       "      <td>1550</td>\n",
       "      <td>328136320</td>\n",
       "      <td>90.22</td>\n",
       "      <td>2287.57</td>\n",
       "      <td>348927115</td>\n",
       "      <td>3.237721e+08</td>\n",
       "      <td>92.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>4.72</td>\n",
       "      <td>4256</td>\n",
       "      <td>294386131</td>\n",
       "      <td>72.12</td>\n",
       "      <td>913.90</td>\n",
       "      <td>347126358</td>\n",
       "      <td>2.804911e+08</td>\n",
       "      <td>80.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>65.27</td>\n",
       "      <td>2337</td>\n",
       "      <td>84756187</td>\n",
       "      <td>59.98</td>\n",
       "      <td>211.02</td>\n",
       "      <td>347126358</td>\n",
       "      <td>2.943582e+07</td>\n",
       "      <td>8.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>18.25</td>\n",
       "      <td>6593</td>\n",
       "      <td>379142318</td>\n",
       "      <td>67.81</td>\n",
       "      <td>693.54</td>\n",
       "      <td>347126358</td>\n",
       "      <td>3.099488e+08</td>\n",
       "      <td>89.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>6.80</td>\n",
       "      <td>6090</td>\n",
       "      <td>157744627</td>\n",
       "      <td>45.38</td>\n",
       "      <td>515.54</td>\n",
       "      <td>347126358</td>\n",
       "      <td>1.470180e+08</td>\n",
       "      <td>42.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>E. coli fragmented</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>4.63</td>\n",
       "      <td>7363</td>\n",
       "      <td>354775080</td>\n",
       "      <td>104.24</td>\n",
       "      <td>440.83</td>\n",
       "      <td>347126358</td>\n",
       "      <td>3.383490e+08</td>\n",
       "      <td>97.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>5.09</td>\n",
       "      <td>10304</td>\n",
       "      <td>101075318</td>\n",
       "      <td>10.25</td>\n",
       "      <td>908.55</td>\n",
       "      <td>117375178</td>\n",
       "      <td>9.593058e+07</td>\n",
       "      <td>81.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>48.78</td>\n",
       "      <td>4512</td>\n",
       "      <td>16982123</td>\n",
       "      <td>10.83</td>\n",
       "      <td>179.06</td>\n",
       "      <td>117375178</td>\n",
       "      <td>8.698243e+06</td>\n",
       "      <td>7.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>11.37</td>\n",
       "      <td>14816</td>\n",
       "      <td>118057441</td>\n",
       "      <td>10.42</td>\n",
       "      <td>677.81</td>\n",
       "      <td>117375178</td>\n",
       "      <td>1.046343e+08</td>\n",
       "      <td>89.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>7.03</td>\n",
       "      <td>4021</td>\n",
       "      <td>33443360</td>\n",
       "      <td>4.23</td>\n",
       "      <td>1444.11</td>\n",
       "      <td>117375178</td>\n",
       "      <td>3.109229e+07</td>\n",
       "      <td>26.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>14.13</td>\n",
       "      <td>7844</td>\n",
       "      <td>115958942</td>\n",
       "      <td>19.22</td>\n",
       "      <td>660.61</td>\n",
       "      <td>117375178</td>\n",
       "      <td>9.957394e+07</td>\n",
       "      <td>84.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Panaroo podstawowy</td>\n",
       "      <td>4.95</td>\n",
       "      <td>3323</td>\n",
       "      <td>40604368</td>\n",
       "      <td>13.64</td>\n",
       "      <td>851.21</td>\n",
       "      <td>45919085</td>\n",
       "      <td>3.859445e+07</td>\n",
       "      <td>84.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Uzupełnienie Panaroo</td>\n",
       "      <td>50.27</td>\n",
       "      <td>2194</td>\n",
       "      <td>22531171</td>\n",
       "      <td>12.58</td>\n",
       "      <td>406.89</td>\n",
       "      <td>45919085</td>\n",
       "      <td>1.120475e+07</td>\n",
       "      <td>24.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Panaroo z uzupełnieniem</td>\n",
       "      <td>21.13</td>\n",
       "      <td>5517</td>\n",
       "      <td>63135539</td>\n",
       "      <td>13.22</td>\n",
       "      <td>683.07</td>\n",
       "      <td>45919085</td>\n",
       "      <td>4.979500e+07</td>\n",
       "      <td>108.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Cactus</td>\n",
       "      <td>3.78</td>\n",
       "      <td>1504</td>\n",
       "      <td>19601305</td>\n",
       "      <td>7.78</td>\n",
       "      <td>1612.74</td>\n",
       "      <td>45919085</td>\n",
       "      <td>1.886038e+07</td>\n",
       "      <td>41.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Pneumococcus</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>9.81</td>\n",
       "      <td>2176</td>\n",
       "      <td>43828686</td>\n",
       "      <td>18.33</td>\n",
       "      <td>991.27</td>\n",
       "      <td>45919085</td>\n",
       "      <td>3.952909e+07</td>\n",
       "      <td>86.08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Dataset                    Model  Udział przerw (%)  \\\n",
       "0                E. coli       Panaroo podstawowy               0.11   \n",
       "1                E. coli     Uzupełnienie Panaroo              62.29   \n",
       "2                E. coli  Panaroo z uzupełnieniem              13.96   \n",
       "3                E. coli                   Cactus               1.48   \n",
       "4   E. coli contaminated       Panaroo podstawowy               0.25   \n",
       "5   E. coli contaminated     Uzupełnienie Panaroo              62.33   \n",
       "6   E. coli contaminated  Panaroo z uzupełnieniem              13.96   \n",
       "7   E. coli contaminated                   Cactus               1.33   \n",
       "8     E. coli fragmented       Panaroo podstawowy               4.72   \n",
       "9     E. coli fragmented     Uzupełnienie Panaroo              65.27   \n",
       "10    E. coli fragmented  Panaroo z uzupełnieniem              18.25   \n",
       "11    E. coli fragmented                   Cactus               6.80   \n",
       "12    E. coli fragmented              Progressive               4.63   \n",
       "13            Klebsiella       Panaroo podstawowy               5.09   \n",
       "14            Klebsiella     Uzupełnienie Panaroo              48.78   \n",
       "15            Klebsiella  Panaroo z uzupełnieniem              11.37   \n",
       "16            Klebsiella                   Cactus               7.03   \n",
       "17            Klebsiella              Progressive              14.13   \n",
       "18          Pneumococcus       Panaroo podstawowy               4.95   \n",
       "19          Pneumococcus     Uzupełnienie Panaroo              50.27   \n",
       "20          Pneumococcus  Panaroo z uzupełnieniem              21.13   \n",
       "21          Pneumococcus                   Cactus               3.78   \n",
       "22          Pneumococcus              Progressive               9.81   \n",
       "\n",
       "    Liczba bloków  Łączna długość zawartych sekwencji  Średni rozmiar bloku  \\\n",
       "0            4234                           292139716                 72.90   \n",
       "1            2246                            83733481                 66.17   \n",
       "2            6480                           375873197                 70.57   \n",
       "3            1333                           332387749                 92.35   \n",
       "4            4247                           292302476                 72.62   \n",
       "5            2251                            82821561                 65.82   \n",
       "6            6498                           375124037                 70.27   \n",
       "7            1550                           328136320                 90.22   \n",
       "8            4256                           294386131                 72.12   \n",
       "9            2337                            84756187                 59.98   \n",
       "10           6593                           379142318                 67.81   \n",
       "11           6090                           157744627                 45.38   \n",
       "12           7363                           354775080                104.24   \n",
       "13          10304                           101075318                 10.25   \n",
       "14           4512                            16982123                 10.83   \n",
       "15          14816                           118057441                 10.42   \n",
       "16           4021                            33443360                  4.23   \n",
       "17           7844                           115958942                 19.22   \n",
       "18           3323                            40604368                 13.64   \n",
       "19           2194                            22531171                 12.58   \n",
       "20           5517                            63135539                 13.22   \n",
       "21           1504                            19601305                  7.78   \n",
       "22           2176                            43828686                 18.33   \n",
       "\n",
       "    Średnia długość sekwencji w bloku  Sumaryczna długość genomów  \\\n",
       "0                              945.35                   347207379   \n",
       "1                              213.45                   347207379   \n",
       "2                              707.48                   347207379   \n",
       "3                             2629.29                   347207379   \n",
       "4                              945.38                   348927115   \n",
       "5                              211.56                   348927115   \n",
       "6                              707.26                   348927115   \n",
       "7                             2287.57                   348927115   \n",
       "8                              913.90                   347126358   \n",
       "9                              211.02                   347126358   \n",
       "10                             693.54                   347126358   \n",
       "11                             515.54                   347126358   \n",
       "12                             440.83                   347126358   \n",
       "13                             908.55                   117375178   \n",
       "14                             179.06                   117375178   \n",
       "15                             677.81                   117375178   \n",
       "16                            1444.11                   117375178   \n",
       "17                             660.61                   117375178   \n",
       "18                             851.21                    45919085   \n",
       "19                             406.89                    45919085   \n",
       "20                             683.07                    45919085   \n",
       "21                            1612.74                    45919085   \n",
       "22                             991.27                    45919085   \n",
       "\n",
       "    Pokrycie pangenomu  Pokrycie pangenomu (%)  \n",
       "0         2.918184e+08                   84.05  \n",
       "1         3.157590e+07                    9.09  \n",
       "2         3.234013e+08                   93.14  \n",
       "3         3.274684e+08                   94.31  \n",
       "4         2.915717e+08                   83.56  \n",
       "5         3.119888e+07                    8.94  \n",
       "6         3.227567e+08                   92.50  \n",
       "7         3.237721e+08                   92.79  \n",
       "8         2.804911e+08                   80.80  \n",
       "9         2.943582e+07                    8.48  \n",
       "10        3.099488e+08                   89.29  \n",
       "11        1.470180e+08                   42.35  \n",
       "12        3.383490e+08                   97.47  \n",
       "13        9.593058e+07                   81.73  \n",
       "14        8.698243e+06                    7.41  \n",
       "15        1.046343e+08                   89.15  \n",
       "16        3.109229e+07                   26.49  \n",
       "17        9.957394e+07                   84.83  \n",
       "18        3.859445e+07                   84.05  \n",
       "19        1.120475e+07                   24.40  \n",
       "20        4.979500e+07                  108.44  \n",
       "21        1.886038e+07                   41.07  \n",
       "22        3.952909e+07                   86.08  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stats"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
